### DISCUSSION TITLE
Is undiscord abandoned/unmaintained?

### DISCUSSION THREAD

**Discussion by skppr** [victornpb/undiscord#446] [1. Questions & Answers] [answered]
This has not been updated since May which was a limited burst of activity from the original author that unfortunately lead some of the best forks to be archived (which oddly was done months after his last update), meaning that now that he has once again gone missing we are left with no choice but to scrap together a working version ourselves from each pull request.

I'm sorry if you think I am lazy or demanding other people work for me for free. I'm just upset because this is the second time he's gone AWOL and this software is in pieces all over github when it doesn't need to be. I am not good at handling software on my own and I'm sure I'm not the only one. As far as I can tell he hasn't said we can't make forks anymore so I don't understand why there isn't a single big and popular one being used. But I'm scared of using less popular forks because of potential malware, as my discord account has a lot of personal conversations.

**Comment by abigailwillow** [comment]


**Reply by CallumADW** [reply]
Can I find the last version of your fork? I wont bug you if it breaks! I read an issue that this one auto adjusts too quick but honestly not sure how to use the fork that someone linked... maybe its easy and im missing something.

**Reply by abigailwillow** [reply]
I've fully discontinued mine. If you find a PR that you would like to use you can generally click the link to the forked repository that the PR is from, then browse to `src/deleteMessages.js` and click raw, which should bring you to a page that looks something like [this](https://raw.githubusercontent.com/0xallie/undiscord/less-aggressive-delay/src/deleteMessages.js) (this specific link is the source code for the fork associated with #419). This code can be used to replace the related code snippet in the installed version of your userscript plugin.

**Reply by CallumADW** [reply]
I'll try that! I was getting confused between src/deleteMessages.js and the really long deleteDiscordMessages.user.js in the main repo.

**Reply by abigailwillow** [reply]
If I'm correct `deleteDiscordMessages.user.js` is built from `src/deleteMessages.js`. Since most PRs won't bother with this step that file will be either the same as the main repository or behind. The files in the `src` directory should be the file that's ahead of the main repository in most cases.

**Reply by CallumADW** [reply]
I just tried what you suggested but tampermonkey says its an invalid userscript. The one in src is 10% the length of deleteDiscordMessages.user.js.... (like 150 lines vs 1.5k) Im usually good with stuff on github but evidently not this one

**Reply by abigailwillow** [reply]
The build step combines the files in the `src` directory, you can't just install the source file it as is anymore. It's unfortunately a little bit more complicated. You will have to edit your currently installed version in Tampermonkey and find where the source code actually starts, at the time of writing it's line [697 up to line 919](https://github.com/victornpb/undiscord/blob/90e66e2319d5fb95612c79c2443303fb6e8f789a/deleteDiscordMessages.user.js#L697-L919), omitting the last line that contains `export default deleteMessages;`. It's cumbersome but it should let you use the code from any PR and even combine if you're competent with this sort of thing.

**Reply by CallumADW** [reply]
alright I just copied the code from `src/deleteMessages.js` and pasted it where that code already existed (unmodified obviously) in `deleteDiscordMessages.user.js` Seems to work!

**Reply by CallumADW** [reply]
We replied at the same time. All good! i appreciate your help.

**Reply by abigailwillow** [reply]
Sorry for the misinformation, I didn't realize this was how it worked initially, I'll update my previous responses to reflect this to make sure someone else coming across this discussion won't get confused.

**Reply by CallumADW** [reply]
No worries, it happens! For some odd reason I can't edit the search or delete delay whilst using the script... but Im not getting rate limited so not worried. If I get limited i'll just hold off for a bit.

**Reply by CallumADW** [reply]
Oh, its meant to be edited using the arrow keys to up and down in intervals. Duh.

**Reply by victornpb** [reply]
> The build step combines the files in the `src` directory, you can't just install the source file it as is anymore. **It's unfortunately a little bit more complicated**. 

I tried simplifying how the build process work last week, hopefully it will be more friendly. Unfortunately the project grown to the point that this kind of things become necessary.

Contributing instructions:
https://github.com/victornpb/undiscord/blob/master/CONTRIBUTING.md

Build tutorial (45 seconds video)
https://www.youtube.com/watch?v=AKTCvzvcPig

**Reply by abigailwillow** [reply]
I totally understand why it's set up this way. Perhaps I used the wrong wording but I didn't intend to say that it's unfortunate that the building is so complicated. Thanks for the little tutorial you posted!

**Comment by victornpb** [comment, marked as answer]
> Is undiscord abandoned/unmaintained?

TLDR, no. 

----

### About the infrequent updates

> This has not been updated since May which was a limited burst of activity from the original author that unfortunately lead some of the best forks to be archived (which oddly was done months after his last update), meaning that now that he has once again gone missing we are left with no choice but to scrap together a working version ourselves from each pull request.

I would like to address a few things, I can only work on this from time to time, and it may seem that it doesn't receive updates as often as some would like, but this project has been an ongoing thing for at least 3 to 4 years. I can't just keep up with all the activity in this repository, because it is very time-consuming, the majority being just noise or just people having general questions (which is fine) but I can't answer each and every one personally. That's the main reason the issues tab has been disabled for some time, (even tho I rather have issues enabled) the discussion format is less developer-centric, and some folks have been able to help answer some questions. (Like @abbydiode, and some other names I will forget to mention, TY)

----
### About PRs

> I'm sorry if you think I am lazy or demanding other people work for me for free. I'm just upset because this is the second time he's gone AWOL and this software is in pieces all over GitHub when it doesn't need to be. I am not good at handling software on my own and I'm sure I'm not the only one.

While there are some valuable contributions from the community from time to time, most are simply not high quality enough. So it's not just a matter of simply clicking "merge" to all of them. I have to read them to understand why, and how it works, make sure it's not malicious, and make sure it fixes one thing and doesn't break another. One buggy contribution and I will wake up the next day with a hundred messages complaining that it broke. Now I have to stop whatever I'm doing to fix it in a hurry.

I've accepted some contributions in the past that were below my quality standards just because it was something that the community wanted. But after a few of those, the code became really messy and hard to modify further.

----

### About forks

> As far as I can tell he hasn't said we can't make forks anymore so I don't understand why there isn't a single big and popular one being used. But I'm scared of using less popular forks because of potential malware, as my discord account has a lot of personal conversations.

Yeah, the code base is fully open-source, so I can't stop anyone from forking it, as long as they respect the license. Just because you can, doesn't mean I want to encourage people to fork it. But rather because I think for a tool like this you need the source to be fully transparent. All my projects are open source, but for this one, I take a bit of a different approach because I think the safety of the users is more important than only caring about the sharing aspect of open-source. I would really like to have more people taking care of this project, but I can't delegate the responsibility I have with people using this.
(Please read my notes about security and forks here https://github.com/victornpb/undiscord/wiki/Security-Policy#note-about-forks)

----

That said being said, I have a full-time job, and I have many other projects (not just public GitHub), recently I suffered an accident and it's been hard to do my main job. But I will still keep working on this as long as it is viable. But that means you just have to be a bit patient sometimes.

I'm working on an [update](https://github.com/victornpb/undiscord/tree/next) tidying things up and hopefully, it will be easier for people to submit _less hacky_ PRs. That new version will contain a few new features (don't expect anything major, at least for that immediate next version), but the main purpose of this version is rewriting and organizing the code. (Unfortunately, that means some PRs are going to be closed because I will not be able to merge them, but like last time I was able to incorporate some of the suggestions myself, the PRs were closed because I couldn't take the PR as is, the reason being the quality as I said before)

I'm also working on semi-automatized testing to make my job of verifying if a PR works correctly easier.

Sincerely,

Victor.

**Reply by abigailwillow** [reply]
It's great to see you shed some light on this. I understand your issue. Managing a full-time job and a private life next to a decently popular open-source project is a lot of work. However to users it does look like an abandoned project whenever you're unable to work on it. Perhaps you could post a similar explanation such as this as a pinned discussion so people are more likely to find it. Could even add a link to it in the readme to improve visibility?

Also I think it'd be valuable if you could find contributors that you can trust that can maintain PRs and issues while you're unable to work on this?

I've been closely following what's been happening to this project for some time now, and what I mainly see is users being bothered by the lack of communication. I think there's a lot of improvement that can be made in that department. 

I'm sorry to hear about your accident and hope you get well soon!

**Reply by victornpb** [reply]
> I've been closely following what's been happening to this project

I'm super grateful for all your contributions @abbydiode, I'm sorry if I didn't manifest this enough.

>  Perhaps you could post a similar explanation such as this as a pinned discussion ...

I can write something short later but for now I pinned this question in the discussion tab.

> I'm sorry to hear about your accident and hope you get well soon!

It's being a slow process, but thank you!

**Comment by victornpb** [comment]
NOTE: The original question was "Is there a most popular fork for this anymore?".
I answered the original question, but I rephrased the question to "Is undiscord abandoned/unmaintained?" to better fit my answer so I can pin it.
